src/lib/Libdb/db_postgres_attr.c

```diff
@@ -91,123 +167,153 @@ struct pg_array {
+
+		attr_resc = NULL;
+		attr_value = NULL;
+		
 		attr_name = val->str;
 		val = (struct str_data *)((char *) val->str + ntohl(val->len));
 
-		attr_value = val->str;
+		attr_flags = val->str;
 		val = (struct str_data *)((char *) val->str + ntohl(val->len));
 
-		if (attr_name) {
-			attrs[i].attr_name[sizeof(attrs[i].attr_name) -1] = '\0';
-			strncpy(attrs[i].attr_name, attr_name, sizeof(attrs[i].attr_name));
-			if ((p = strchr(attrs[i].attr_name, '.'))) {
-				*p = '\0';
-				attrs[i].attr_resc[sizeof(attrs[i].attr_resc) -1] = '\0';
-				strncpy(attrs[i].attr_resc, p + 1, sizeof(attrs[i].attr_resc));
-			} else
-				attrs[i].attr_resc[0] = '\0';
-		} else
-			attrs[i].attr_name[0] = 0;
-
-		if (attr_value && (p = strchr(attr_value, '.'))) {
-			*p ='\0';
-			attrs[i].attr_flags = atol(attr_value);
-			attrs[i].attr_value = strdup(p + 1);
-		} else {
-			attrs[i].attr_flags = 0;
-			attrs[i].attr_value = NULL;
+		if ((attr_resc = strchr(attr_name, '.'))) {
+			*attr_resc = '\0';
+			attr_resc++;
 		}
+
+		if ((attr_value = strchr(attr_flags, '.'))) {
+			*attr_value = '\0';
+			attr_value++;
+		}
+
+		flags = strtol(attr_flags, &endp, 10);
+		if (*endp != '\0')
+			return -1;
+
+		if (!(pal = make_attr(attr_name, attr_resc, attr_value, flags)))
+			return -1;
+		
+		append_link(&(attr_list->attrs), &pal->al_link, pal);
 	}
 	attr_list->attr_count = i;
 	return 0;
 }
 
 /**
  * @brief
- *	Converts an attribute list to string array which is in the form of postgres hstore.
+ *	Converts an PBS link list of attributes to DB hstore(array) format
  *
- * @param[in]	attr_list - List of pbs_db_attr_list_t objects
  * @param[out]  raw_array - Array string which is in the form of postgres hstore
+ * @param[in]	attr_list - List of pbs_db_attr_list_t objects
+ * @param[in]	keys_only - if true, convert only the keys, not values also
  *
  * @return      Error code
  * @retval	-1 - On Error
  * @retval	 0 - On Success
  *
  */
 int
-convert_db_attr_list_to_array(char **raw_array, pbs_db_attr_list_t *attr_list)
+attrlist_2_dbarray_ex(char **raw_array, pbs_db_attr_list_t *attr_list, int keys_only)
 {
-	int i;
-	pbs_db_attr_info_t *attrs = attr_list->attributes;
 	struct pg_array *array;
 	int len = 0;
 	struct str_data *val = NULL;
 	int attr_val_len = 0;
+	svrattrl *pal;
+	int count = 0;
 
 	len = sizeof(struct pg_array);
-	for (i = 0; i < attr_list->attr_count; i++) {
+	for (pal = (svrattrl *)GET_NEXT(attr_list->attrs); pal != NULL; pal = (svrattrl *)GET_NEXT(pal->al_link)) {
 		len += sizeof(int32_t) + PBS_MAXATTRNAME + PBS_MAXATTRRESC + 1; /* include space for dot */
-		attr_val_len = (attr_list->attributes[i].attr_value == NULL? 0:strlen(attr_list->attributes[i].attr_value));
+		if (keys_only)
+			attr_val_len = 0;
+		else
+			attr_val_len = (pal->al_atopl.value == NULL? 0:strlen(pal->al_atopl.value));
 		len += sizeof(int32_t) + 3 + attr_val_len + 1; /* include space for dot */
+		count++;
 	}
 
+	assert(count == attr_list->attr_count);
```

Comment:
Yes, sure - asserts are bad in general. But this would have been a programming mistake, so is a debug thing only and unlikely to ever trigger. However, since we do not turn on NDEBUG, they could. I converted to a return -1